song-python-sdk/overture_song/validation.py

```diff
@@ -0,0 +1,176 @@
+                            "The required list datafield '{}' was supplied an empty array",
+                            self.name)
+            if value is not None:
+                for t in self.types:
+                    for item in value:
+                        self._validate_single(t, item)
+        else:
+            for t in self.types:
+                self._validate_single(t, value)
+
+    def _validate_single(self, t, value):
+        if self.required:
+            check_state(value is not None, "The datafield '{}' is required", self.name)
+
+        if value is not None:
+            if self.required:
+                if t is str:
+                    check_state(value, "The required string datafield '{}' was supplied an empty string ",
+                                self.name)
+            check_state(isinstance(value, t),
+                        "The datafield '{}' is of '{}' type, but was supplied a value of type '{}' with value '{}'",
+                        self.name, t, type(value), value)
+
+
+"""
+    Validation decorator that intercepts "setting" of properties/attributes defined by the @dataclass decorator. 
+    Once intercepted, it validates the input value to set against predefined rules, encapsulated in a DataField class
+"""
+
+
+# TODO: Update implementation to process dataclass definition. Should extract properties, their types, if they are a
+# list or not, if they are required or optional
+class validation(object):
+    def __init__(self, *datafields):
+        self.datafields = list(datafields)
+        utils.check_type(self.datafields, list)
+        check_state(len(self.datafields) > 0, "Must define atleast one datafield")
+        self.name_type_map = {}
+        for datafield in self.datafields:
+            utils.check_type(datafield, DataField)
+            if datafield.name not in self.name_type_map:
+                self.name_type_map[datafield.name] = {}
+            for t in datafield.types:
+                t_name = t.__name__
+
+                if t_name in self.name_type_map[datafield.name]:
+                    raise Exception(
+                        "Collision: The datafield definition '{}' already exists as '{}' for the type '{}'".format(
+                            datafield, self.name_type_map[datafield.name][t_name], t_name))
+                self.name_type_map[datafield.name][t_name] = datafield
+
+    def __call__(self, Cls):
+        name_type_map = self.name_type_map
+        datafields = self.datafields
+
+        class Validator(Cls):
+
+            SPECIAL_FIELD = '__dataclass_fields__'
+            _INTERNAL_DICT = Cls.__dict__[SPECIAL_FIELD]
+
+            def __init__(self, *args, **kwargs):
+
+                Cls.__init__(self, *args, **kwargs)
+                check_state(Validator.SPECIAL_FIELD in Cls.__dict__,
+                            "Decorator can only process dataclasses")
+                self._check_validator()
+
+            def _check_validator(self):
```

Comment:
This guy can be made a static method or just a function. It doesn't make use of the object instance through `self`. 

```python
@staticmethod
def _check_validator():
```